Manage affinity groups with cloudstack provider

schans opened the pull request:
Hi all!

Please find my pull request for managing affinity groups in cloudstack. It would be great if somebody could have a look at it and provide feedback or merge. I'm happy to work on this further if needed.

Best regards,
Bart

schans commented:
Hi,

I fixed a bug when destroying affinity groups. The commits have been added to the pull request.

Bart

schans commented:
Thanks for adding the labels!

It would be really awesome if this could make it in one of the next 0.7.x releases. Anything I can do to help make that possible?

Bart

svanharmelen commented on builtin/providers/cloudstack/resource_cloudstack_affinity_group.go:
```diff
@@ -0,0 +1,136 @@
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
```
For consistency with the other resources, could you change the order to `name`, `description`, `type, `project`?

svanharmelen commented on builtin/providers/cloudstack/resource_cloudstack_affinity_group.go:
```diff
@@ -0,0 +1,136 @@
+	p := cs.AffinityGroup.NewCreateAffinityGroupParams(name, affinityGroupType)
+
+	// Set the description
+	description, hasDescription := d.GetOk("description")
```
It's more idiomatic to write this as: `description, ok := d.GetOk("description"`. So to keep things consistent with the existing code, please also use that here.

svanharmelen commented on vendor/github.com/xanzy/go-cloudstack/cloudstack/AffinityGroupService.go:
```diff
@@ -98,7 +98,7 @@ func (p *CreateAffinityGroupParams) SetType(v string) {
 		p.p = make(map[string]interface{})
 	}
-	p.p["affinityGroupType"] = v
+	p.p["type"] = v
```
Hmm, this doesn't look good...

You should never change a vendored package directly. This would mean that we now effectively forked this package, without it being documented anywhere.

If this is somethings that needs to be fixed in the `go-cloudstack` package, it should be fixed in there (at `github.com/xanzy/go-cloudstack`). And then after that we can update the vendored package in here to pull in the fix.

svanharmelen commented on vendor/github.com/xanzy/go-cloudstack/cloudstack/AffinityGroupService.go:
```diff
@@ -108,7 +108,7 @@ func (s *AffinityGroupService) NewCreateAffinityGroupParams(name string, affinit
 	p.p = make(map[string]interface{})
 	p.p["name"] = name
-	p.p["affinityGroupType"] = affinityGroupType
+	p.p["type"] = affinityGroupType
```
Of course the same goes for this one...

svanharmelen commented on website/source/docs/providers/cloudstack/r/affinity_group.html.markdown:
```diff
@@ -0,0 +1,43 @@
+The following attributes are exported:
+
+* `id` - The id of the affinity group.
+* `name` - The name of the affinity group.
```
In order for this new file to be shown in the index, it should also be added here: https://github.com/hashicorp/terraform/blob/master/website/source/layouts/cloudstack.erb

svanharmelen replied:
The `name` attribute should not be listed here, as this is already listed above. Only computed attributes should be listed here in addition to the internal `id` attribute.

svanharmelen commented on website/source/docs/providers/cloudstack/r/affinity_group.html.markdown:
```diff
@@ -0,0 +1,43 @@
+* `description` - (Optional) The description of the affinity group.
+
+* `project` - (Optional) The name or ID of the project to register this
+    affinity group to. Changing this forces a new resource to be created.
```
Again for consistency, please update the order to `name`, `description`, `type`, `project`.

svanharmelen commented on website/source/docs/providers/cloudstack/r/affinity_group.html.markdown:
````diff
@@ -0,0 +1,43 @@
+```
+resource "cloudstack_affinity_group" "default" {
+  name = "myGroup"
+  type = "host anti-affinity"
````
Is this a valid type? I thought only `affinity` and `anti-affinity` were valid types?

svanharmelen commented on builtin/providers/cloudstack/resource_cloudstack_affinity_group.go:
```diff
@@ -0,0 +1,136 @@
+		return err
+	}
+
+	log.Printf("[DEBUG] New affinity group successfully generated at Cloudstack")
```
_nitting_ Maybe this looks cleaner: `log.Printf("[DEBUG] New affinity group successfully created")`

svanharmelen commented on builtin/providers/cloudstack/resource_cloudstack_affinity_group.go:
```diff
@@ -0,0 +1,136 @@
+		log.Printf("[DEBUG] Affinity group %s does not exist", d.Id())
+		d.SetId("")
+		return nil
+	}
```
This whole part (line 83 through 99) could/should be written as:

```go
// Get the affinity group details
ag, count, err := cs.AffinityGroup.GetAffinityGroupByID(
  d.Id(),
  cloudstack.WithProject(d.Get("project").(string)),
)
if err != nil {
  if count == 0 {
    log.Printf("[DEBUG] Affinity group %s does not longer exist", d.Id())
    d.SetId("")
    return nil
  }

  return err
}
```

svanharmelen commented on builtin/providers/cloudstack/resource_cloudstack_affinity_group.go:
```diff
@@ -0,0 +1,136 @@
+	//Affinity group name is unique in a cloudstack account so dont need to check for multiple
+	d.Set("name", r.AffinityGroups[0].Name)
+	d.Set("description", r.AffinityGroups[0].Description)
+	d.Set("type", r.AffinityGroups[0].Type)
```
This should then also be changed slightly as there is just one group returned.

svanharmelen commented on builtin/providers/cloudstack/resource_cloudstack_affinity_group.go:
```diff
@@ -0,0 +1,136 @@
+		// This is a very poor way to be told the ID does no longer exist :(
+		if strings.Contains(err.Error(), fmt.Sprintf(
+			"Unable to find affinity group %s", d.Id())) {
+			return nil
```
Does the API for this call differ from the other API responses? As the more generic way to check for this (which is used in all other places) is:

```go
if strings.Contains(err.Error(), fmt.Sprintf(
  "Invalid parameter id value=%s due to incorrect long value format, "+
    "or entity does not exist", d.Id())) {
  return nil
}
```

svanharmelen commented:
Thanks for this @schans! Even through I have some comments, this looks pretty good and complete 👍 

Most comments are easy and small to fix/improve, but the changes to the vendored package need to be handled properly before we can merge this one.

Please ping me if you need help with that...

schans commented:
Hi @svanharmelen,

Thank you very much for reviewing! I'll make an update as soon as possible and try to find a way to work around the "bug" in the xanzy/go-cloudstack for the time being and if needed create a patch for that as well. 

Bart

svanharmelen commented:
@schans sounds good! Yet I don't think you can work around the problem in the `go-cloudstack` package, that will need to get fixed in order for you to be able to use this correctly.,

But I can pick that part up, will have a look at it shortly...

svanharmelen commented:
@schans that was actually a nice bug, but I fixed it already: https://github.com/xanzy/go-cloudstack/pull/60

So that means we (you, or me if you would like me to fix that) should update the vendored package in here to pull the fix in.

schans commented:
@svanharmelen, I have implemented your feedback. Thanks again! 

Now looking into updating the vendored go-cloudstack package to v2.1.2 release.

schans commented:
@svanharmelen, I have updated the go-cloudstack package to v2.1.2 in a523f4c. For now I ignored the returned count in resources.go L50-60, 95 since an error is returned when there is no exact match.

Bart

schans commented:
@svanharmelen, I'm also running some integration tests here and I'm getting some odd errors that more than one result is found for (some) affinity groups. Investigating...

svanharmelen commented:
Check... Then I'll do nothing till your done 😀

schans commented:
Ok, did some more debugging. Strangely enough the Count variable in the list result doesn't seem to be reliable. I guess it might be returned wrongly by the api? I added some debugging locally to print when the l.Count is not equal to the actual length of the l.AffinityGroups slice:

2016/08/29 18:04:32 [DEBUG] plugin: terraform: cloudstack-provider (internal) 2016/08/29 18:04:32 [ERROR] Count not equal to actual length 3 vs 1
2016/08/29 18:04:32 [DEBUG] plugin: terraform: cloudstack-provider (internal) 2016/08/29 18:04:32 [ERROR] Count not equal to actual length 2 vs 1
2016/08/29 18:04:32 [DEBUG] plugin: terraform: cloudstack-provider (internal) 2016/08/29 18:04:32 [ERROR] Count not equal to actual length 2 vs 1

So we do get a single response result from the api when the id filter is set, but the count can be "something else". 

I'll implement a workaround for this since I'm not sure if this a bug in the vendor library or something with Cloudstack itself. The go-cloudstack library does have the comment: "This is a courtesy helper function, which in some cases may not work as expected!". This might be one of those cases?

schans commented:
Implemented the workaround in 3e97427. Basically just get the list and use the length of the slice as count instead of the provided one. Still not sure why the count is set wrongly, but it does break the GetAffinityGroupByID method.

svanharmelen commented:
@schans please delete your last commit with the workaround. Let's not try to work around issues, but let's try to solve them directly so we don't create any unneeded technical dept for ourselves.

> The go-cloudstack library does have the comment: "This is a courtesy helper function, which in some cases may not work as expected!". This might be one of those cases?

Nope, this is not one of those cases. Both the count and the results are returned by the CloudStack API and not altered by the `go-cloudstack` package. So it looks like this is a bug in the CloudStack API.

Let us take a brief moment to check what is going on and if needed file a bug with CloudStack. If it does turn out to be a bug in the CloudStack API, I can update the helper functions in the `go-cloudstack` package so they discard the count and take the length of the result instead. That is an easy and quick fix that would make it work for all Go programs using the `go-cloudstack` package.

But first some questions... Are you familiar with `cloudmonkey`? Could [set display json](https://cwiki.apache.org/confluence/display/CLOUDSTACK/CloudStack+cloudmonkey+CLI#CloudStackcloudmonkeyCLI-JSONoutput) and make some list calls to see what the output is?

```
> set display json
> list affinitygroups id=xxxxx
```

And could you tell me what version of CloudStack you are running?

Thanks!

schans commented:
Hi @svanharmelen!

Thanks again for helping out! I was indeed probably a bit over-practical with my "workaround" ;-) I'll revert that one. The terraform version we are working against is 4.7. From the cloudmonkey output (would be really nice if go-cloudstack could also produce this kind of output, but that is for another time):

With one result (aka correct):

`
> list affinitygroups id=96ce22ab-8570-408f-b90c-5a45af340075
count = 1
affinitygroup:
id = 96ce22ab-8570-408f-b90c-5a45af340075
name = nameserver
description = Nameserver host anti-affinity
domain = *****
domainid = 7af83bfc-834b-4c5e-963b-74dee0418972
project = *****
projectid = ebd9c9a3-68f7-4d46-a2f1-a8ad1e46f9f4
type = host anti-affinity
virtualmachineIds:
{u'count': 1, u'affinitygroup': [{u'domain': *****', u'domainid': u'7af83bfc-834b-4c5e-963b-74dee0418972', u'description': u'Nameserver host anti-affinity', u'virtualmachineIds': [u'539cac51-5aef-4e33-8e3e-57f9d0b1cc82'], u'projectid': u'ebd9c9a3-68f7-4d46-a2f1-a8ad1e46f9f4', u'project': u'*******', u'type': u'host anti-affinity', u'id': u'96ce22ab-8570-408f-b90c-5a45af340075', u'name': u'nameserver'}]}
`

With result of count=2:
`
> list affinitygroups id=2fb3568e-bc2c-4d9c-a7d7-1b3ba74c1378
count = 2
affinitygroup:
id = 2fb3568e-bc2c-4d9c-a7d7-1b3ba74c1378
name = test
description = Test host anti-affinity
domain = *******
domainid = 7af83bfc-834b-4c5e-963b-74dee0418972
project = *******
projectid = ebd9c9a3-68f7-4d46-a2f1-a8ad1e46f9f4
type = host anti-affinity
virtualmachineIds:
{u'count': 2, u'affinitygroup': [{u'domain': ******', u'domainid': u'7af83bfc-834b-4c5e-963b-74dee0418972', u'description': u'Test host anti-affinity', u'virtualmachineIds': [u'f6e123fc-7dc2-47ce-90b0-b1ccb3f25f08', u'f9b610ef-daf8-40a9-86bd-0d18fc022509'], u'projectid': u'ebd9c9a3-68f7-4d46-a2f1-a8ad1e46f9f4', u'project': u'*****', u'type': u'host anti-affinity', u'id': u'2fb3568e-bc2c-4d9c-a7d7-1b3ba74c1378', u'name': u'test'}]}
================================================================================
{u'count': 2, u'affinitygroup': [{u'domain': *****', u'domainid': u'7af83bfc-834b-4c5e-963b-74dee0418972', u'description': u'Test host anti-affinity', u'virtualmachineIds': [u'f6e123fc-7dc2-47ce-90b0-b1ccb3f25f08', u'f9b610ef-daf8-40a9-86bd-0d18fc022509'], u'projectid': u'ebd9c9a3-68f7-4d46-a2f1-a8ad1e46f9f4', u'project': u'*****', u'type': u'host anti-affinity', u'id': u'2fb3568e-bc2c-4d9c-a7d7-1b3ba74c1378', u'name': u'test'}]}
`

With result of count=3:
`
> list affinitygroups id=4a681a40-9c73-4ba9-91f4-23450bae4363
count = 3
affinitygroup:
id = 4a681a40-9c73-4ba9-91f4-23450bae4363
name = database
description = Database host anti-affinity
domain = *****
domainid = 7af83bfc-834b-4c5e-963b-74dee0418972
project = ******
projectid = ebd9c9a3-68f7-4d46-a2f1-a8ad1e46f9f4
type = host anti-affinity
virtualmachineIds:
{u'count': 3, u'affinitygroup': [{u'domain': u'******', u'domainid': u'7af83bfc-834b-4c5e-963b-74dee0418972', u'description': u'Database host anti-affinity', u'virtualmachineIds': [u'67ffd6b1-3c4f-4fab-897b-49dfc5155058', u'4f7d0502-d5a4-40fb-b185-4f4e2fd78d78', u'945e7d40-d883-4b89-9e86-7b054a7be555'], u'projectid': u'ebd9c9a3-68f7-4d46-a2f1-a8ad1e46f9f4', u'project': u'******', u'type': u'host anti-affinity', u'id': u'4a681a40-9c73-4ba9-91f4-23450bae4363', u'name': u'database'}]}
================================================================================
{u'count': 3, u'affinitygroup': [{u'domain': u'******', u'domainid': u'7af83bfc-834b-4c5e-963b-74dee0418972', u'description': u'Database host anti-affinity', u'virtualmachineIds': [u'67ffd6b1-3c4f-4fab-897b-49dfc5155058', u'4f7d0502-d5a4-40fb-b185-4f4e2fd78d78', u'945e7d40-d883-4b89-9e86-7b054a7be555'], u'projectid': u'ebd9c9a3-68f7-4d46-a2f1-a8ad1e46f9f4', u'project': u'******', u'type': u'host anti-affinity', u'id': u'4a681a40-9c73-4ba9-91f4-23450bae4363', u'name': u'database'}]}
================================================================================
{u'count': 3, u'affinitygroup': [{u'domain': u'******', u'domainid': u'7af83bfc-834b-4c5e-963b-74dee0418972', u'description': u'Database host anti-affinity', u'virtualmachineIds': [u'67ffd6b1-3c4f-4fab-897b-49dfc5155058', u'4f7d0502-d5a4-40fb-b185-4f4e2fd78d78', u'945e7d40-d883-4b89-9e86-7b054a7be555'], u'projectid': u'ebd9c9a3-68f7-4d46-a2f1-a8ad1e46f9f4', u'project': u'******', u'type': u'host anti-affinity', u'id': u'4a681a40-9c73-4ba9-91f4-23450bae4363', u'name': u'database'}]}
`

So it looks like an issue with the cloudstack api. For each virtual machine that uses the affinity group an identical result is returned. 

I do not have access to other Cloudstack platforms but it might be interesting to check if this behavior is 4.7 specific or even this install specific.

Bart

schans commented:
@svanharmelen , I have squashed the commits into two commits:
- b9abd9a : terraform affinity group code
- 19da7d9 : update vendor dep to go-cloudstack v2.1.2

Please let me know if that is what you intended.

svanharmelen commented:
@schans thanks for this valuable info and for squashing the commits! 👍 

I will take this to our CloudStack developers to get more background/insights about this one and see how to properly solve or workaround this... Will get back to you asap.

svanharmelen commented:
@schans it indeed seems to be a bug in the CloudStack API. I cannot reproduce the behaviour you saw with getting multiple responses, but it does wrongly report the number of VM's in the group as the number of groups found/returned.

I "fixed" this in the `go-cloudstack` package, so if you can update the vendored dependency once again things should go a lot better now.

Please also have another look at your commits as it seems the squashing didn't work as expected. Let me know if you need any help with that!

Once you're done updating/testing/squashing please give me a ping so I can do a last look over the PR  and verify that all the tests pass. After that we can move ahead and merge this baby 😀 

Thx!

schans commented:
Hi @svanharmelen, thanks for implementing the fix/workaround in go-cloudstack! I have updated the vendor dependency, ran some tests successfully and squashed the commits into a single one in e2f8c8d. I have to admit that the squashing is still a bit new to me ;-) 

As far as I can tell everything looks ready to go! 👍   

Travis seems to have some unrelated timeout issues at the moment.

svanharmelen commented:
@schans looks great! Thx!

Will have one final look and also run some tests tomorrow morning...

svanharmelen commented:
@schans it seems the new acceptance tests for this new resource are failing. All existing acceptance tests still pass with the updated `go-cloudstack` package, so that's good news.

In order to move this one forward, I'll go ahead and merge this one as is. Right after that I will make an additional small PR to fix the acceptance tests and add some minor cosmetic tweaks.

Thanks again for your PR and for your help in finding and fixing a couple of issues in the `go-cloudstack` package! Makes the general support for `CloudStack` a bit better again :wink:

svanharmelen commented:
The last few fixes and tweaks are done in PR #8598 So all is good now and we have a nice new `CloudStack` resource... 😀

schans commented:
Awesome!  Thanks for your support and patience!
